frontend: Use qobject_cast when casting QObjects

cg2121 opened the pull request:
### Description
When casting between QObjects, qobject_cast should always be used. This increases performance as there is no RTTI (Run Time Type Information) with qobject_cast, like there is with dynamic_cast.

Using reinterpret_cast is bad practice, as you should use it only in very specific cases.

### Motivation and Context
Better code

### How Has This Been Tested?
compiled and ran OBS

### Types of changes
- Bug fix (non-breaking change which fixes an issue) -->
- Performance enhancement (non-breaking change which improves efficiency)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

cg2121 commented:
Converting to draft, as #11785 needs to be merged first.

cg2121 commented:
This PR now builds correctly

PatTheMav reviewed (approved):
Overall I'm fine with this change, though it obviously touches a whole lot casts in the code that touch a whole lot of aspects of our frontend code, so this might be too late for the next version but should possibly merged after the next version is stabilised to give it some time to exist on the master branch.

@RytoEX WDYT?

sebastian-s-beckmann reviewed (changes requested):
Just looking at it, I think most of these should use static_cast instead of either dynamic_cast or qobject_cast. qobject_cast is better than dynamic_cast if you want the dynamic behavior (i.e. have the object be null if the cast fails), but if you already know/assume that the cast will succeed (and dont check for new_obj == nullptr), you can just use static_cast.

Also, original git commit authorship appears to have gotten lost, that should probably be reverted.

RytoEX commented:
This requires a rebase.

Warchamp7 commented:
> This requires a rebase.

Rebased and conflicts resolved, however I've also added a to-be-squashed commit with some additional casts that were either introduced in the rebase, or missed originally.

sebastian-s-beckmann commented on frontend/settings/OBSBasicSettings.cpp:
```diff
@@ -1216,7 +1216,7 @@ void OBSBasicSettings::LoadBranchesList()
 		// Disable item if branch is disabled
 		if (!branch.is_enabled) {
-			QStandardItemModel *model = dynamic_cast<QStandardItemModel *>(ui->updateChannelBox->model());
+			QStandardItemModel *model = qobject_cast<QStandardItemModel *>(ui->updateChannelBox->model());
```
`static_cast` (also below)

sebastian-s-beckmann commented on frontend/components/VolumeControl.cpp:
```diff
@@ -505,7 +505,7 @@ void VolumeControl::showVolumeControlMenu(QPoint pos)
 void VolumeControl::renameSource()
 {
-	QAction *action = reinterpret_cast<QAction *>(sender());
+	QAction *action = qobject_cast<QAction *>(sender());
```
`static_cast`

Warchamp7 replied:
Added a guard instead since reliance on `sender()` being correct is not ideal.

sebastian-s-beckmann commented on frontend/dialogs/OBSBasicSourceSelect.cpp:
```diff
@@ -763,7 +763,7 @@ void OBSBasicSourceSelect::createNew()
 	OBSSceneItem item = addResult.value();
 
-	OBSBasic *main = reinterpret_cast<OBSBasic *>(App()->GetMainWindow());
+	OBSBasic *main = qobject_cast<OBSBasic *>(App()->GetMainWindow());
```
`OBSBasic::Get()`

sebastian-s-beckmann commented on frontend/importer/ImporterEntryPathItemDelegate.cpp:
```diff
@@ -154,7 +154,7 @@ void ImporterEntryPathItemDelegate::handleClear(QWidget *container)
 void ImporterEntryPathItemDelegate::updateText()
 {
-	QLineEdit *lineEdit = dynamic_cast<QLineEdit *>(sender());
+	QLineEdit *lineEdit = qobject_cast<QLineEdit *>(sender());
```
`static_cast`

sebastian-s-beckmann commented on frontend/utility/audio-encoders.cpp:
```diff
@@ -65,7 +65,7 @@ static void HandleSampleRate(obs_property_t *prop, const char *id)
-	auto main = reinterpret_cast<OBSMainWindow *>(App()->GetMainWindow());
+	auto main = qobject_cast<OBSMainWindow *>(App()->GetMainWindow());
 	if (!main) {
 		blog(LOG_ERROR, "Failed to get main window while populating "
```
`OBSBasic::Get()`, and remove the error handling?

sebastian-s-beckmann commented on frontend/widgets/OBSBasic.cpp:
```diff
@@ -2105,7 +2105,7 @@ void OBSBasic::UpdateTitleBar()
 OBSBasic *OBSBasic::Get()
 {
-	return reinterpret_cast<OBSBasic *>(App()->GetMainWindow());
+	return qobject_cast<OBSBasic *>(App()->GetMainWindow());
```
`static_cast`

sebastian-s-beckmann commented on frontend/widgets/OBSBasic_Clipboard.cpp:
```diff
@@ -187,7 +187,7 @@ void OBSBasic::SourcePasteFilters(OBSSource source, OBSSource dstSource)
 void OBSBasic::actionCopyFilters()
 {
-	QAction *action = reinterpret_cast<QAction *>(sender());
+	QAction *action = qobject_cast<QAction *>(sender());
```
`static_cast` (also below)

Warchamp7 replied:
Added a guard instead since reliance on `sender()` being correct is not ideal.

sebastian-s-beckmann commented on frontend/widgets/OBSBasic_ContextToolbar.cpp:
```diff
@@ -145,10 +145,10 @@ void OBSBasic::UpdateContextBar(bool force)
 		QLayoutItem *la = ui->emptySpace->layout()->itemAt(0);
 		if (la) {
-			if (SourceToolbar *toolbar = dynamic_cast<SourceToolbar *>(la->widget())) {
+			if (SourceToolbar *toolbar = qobject_cast<SourceToolbar *>(la->widget())) {
```
`static_cast` (both)

sebastian-s-beckmann reviewed (changes requested):
I haven't gone through all the files (only about one third) but the idea should be clear.
There are still a lot of `qobject_cast`s that should be `static_cast`s. `qobject_cast` is only needed where it isn't clear (or assumed) that the cast succeeds. Especially in cases where the pointer is dereferenced immediately after the cast, `qobject_cast` doesn't make sense. Also in some cases, instead of casting from the main window it should just be `OBSBasic::Get()`.
Also commit authorship is still changed.

Warchamp7 commented:
> I haven't gone through all the files (only about one third) but the idea should be clear. There are still a lot of `qobject_cast`s that should be `static_cast`s. `qobject_cast` is only needed where it isn't clear (or assumed) that the cast succeeds. Especially in cases where the pointer is dereferenced immediately after the cast, `qobject_cast` doesn't make sense. Also in some cases, instead of casting from the main window it should just be `OBSBasic::Get()`. Also commit authorship is still changed.

Commit authorship is intentional since at this point it contains changes from both of us.

Will address all the feedback.

Warchamp7 committed:
frontend: Avoid reinterpret_cast when casting QObjects

When casting between QObjects, qobject_cast should always be used for dynamic
casts. This increases performance as there is no RTTI (Run Time Type
Information) with qobject_cast, like there is with dynamic_cast. In other cases
where we are sure or assume the cast is valid, use static_cast.

Using reinterpret_cast is bad practice, as you should use it in very
specific cases.

Co-Authored-By: Clayton Groeneveld <19962531+cg2121@users.noreply.github.com>

sebastian-s-beckmann commented on shared/properties-view/properties-view.cpp:
```diff
@@ -1909,7 +1909,7 @@ void WidgetInfo::GroupChanged(const char *setting)
 {
 	const char *setting = obs_property_name(property);
-	QListWidget *list = reinterpret_cast<QListWidget *>(widget);
+	QListWidget *list = qobject_cast<QListWidget *>(widget);
```
`static_cast` (also all other in file, I think).

Warchamp7 replied:
Hopefully got all of them now.

sebastian-s-beckmann reviewed (approved):
Seems correct

PatTheMav commented on frontend/widgets/OBSBasic_ContextToolbar.cpp:
```diff
@@ -145,10 +145,10 @@ void OBSBasic::UpdateContextBar(bool force)
 		QLayoutItem *la = ui->emptySpace->layout()->itemAt(0);
 		if (la) {
-			if (SourceToolbar *toolbar = dynamic_cast<SourceToolbar *>(la->widget())) {
+			if (SourceToolbar *toolbar = static_cast<SourceToolbar *>(la->widget())) {
```
This change crashes OBS with a media source because the cast always succeeds and then `toolbar->GetSource()` will interpret arbitrary memory in the `MediaControls` object as an `OBSWeakSource` and crash deep within `libobs`'s reference counting system.

As can be seen below, this entire code depends on this cast failing if the widget is indeed a `MediaSource` and not a `SourceToolbar` (hence the branch below), so `dynamic_cast` or `qobject_cast` is necessary here.

Warchamp7 replied:
This one never should have been changed to static_cast since it has a conditional check on it. Will fix.

PatTheMav commented on frontend/dialogs/OBSYoutubeActions.cpp:
```diff
@@ -20,7 +20,7 @@ const QString IndexOfGamingCategory = "20";
 	: QDialog(parent),
 	  ui(new Ui::OBSYoutubeActions),
-	  apiYouTube(dynamic_cast<YoutubeApiWrappers *>(auth)),
+	  apiYouTube(static_cast<YoutubeApiWrappers *>(auth)),
```
This will introduce a potential crash if a `TwitchAuth` or any other object with the common `Auth` base class is passed into the constructor.

I'd prefer to keep the `dynamic_cast` here and add a `TODO` to refactor this (and open a new GitHub issue to keep track).

IMO the constructor should require a `YoutubeApiWrappers` object here and should not even allow a base class object to be passed in (it doesn't really make sense from an API design point of view).

It's the _caller_'s responsibility to properly upcast the generic `Auth` object `OBSBasic` provides into the `YoutubeApiWrappers` pointer and potentially abort if that isn't possible.

PatTheMav commented on frontend/docks/YouTubeAppDock.cpp:
```diff
@@ -381,7 +381,7 @@ YoutubeApiWrappers *YouTubeAppDock::GetYTApi()
 	Auth *auth = OBSBasic::Get()->GetAuth();
 	if (auth) {
-		YoutubeApiWrappers *apiYouTube(dynamic_cast<YoutubeApiWrappers *>(auth));
+		YoutubeApiWrappers *apiYouTube(static_cast<YoutubeApiWrappers *>(auth));
```
I'm unsure about this, as `Auth` can be any other derived auth object apart from `YoutubeApiWrappers` which just inherits from `Auth` and adds a whole lot of additional functionality.

Indeed this whole function's purpose is to return the "API object" (so probably a whole lot of stuff other than OAuth-related functionality), so it needs to fail if the object in memory cannot be cast into `YoutubeApiWrappers` (and otherwise the `if (apiYouTube)` conditional doesn't make much sense).

Once again I'd probably opt for keeping `dynamic_cast` here and add a `TODO` for refactoring (probably decoupling the API implementation from the OAuth implementation to make this safer).

PatTheMav commented on frontend/importer/ImporterEntryPathItemDelegate.cpp:
```diff
@@ -154,7 +154,7 @@ void ImporterEntryPathItemDelegate::handleClear(QWidget *container)
 void ImporterEntryPathItemDelegate::updateText()
 {
-	QLineEdit *lineEdit = dynamic_cast<QLineEdit *>(sender());
+	QLineEdit *lineEdit = static_cast<QLineEdit *>(sender());
```
Is there value to add a small comment here (and throughout this PR) to explain and point out _why_ the type can be assumed safely, referencing the base types in an associated UI file or where the associated widget is initialised?

In cases like these it would be helpful to establish the logical connection between this event handler and the code that sets up the handler, in a way pointing out that

> if _that_ code over there is changed, _this_ code over here will crash

which might otherwise get lost in the thousands of lines of code we have.

This would apply to all correct uses of `static_cast` throughout, like:

* Casting to a specific layout type from generic `QLayout` - this creates an implicit link between the UI file (or explicit UI setup code) creating and setting the specific layout sub-type and the `static_cast`.
* Casting to a specific model type from `QAbstractItemModel` - `QWidget`s have a `setModel` method that allow new models to be set, so if any API method outside of the abstract item model is used, why can we be "sure" that it's the specific item model we're casting to (and are we sure that nobody else calls `setModel` ever?).
* Casting any generic `QWidget` to a more specific type - similar situation to the cast to a specific `QLayout`.

The main purpose would be to create a paper trail for these connections that are non-obvious in code, as the compiler will not be able to catch these mistakes.

Warchamp7 replied:
> Is there value to add a small comment here

Ehhh. In the purest sense, yes. But in this example, and in most if not all of the other ones mentioned, I believe it is a mistake in the code itself that we should be fixing.

Usage of sender() and receiver() is a legacy approach from before Qt slots could use lambdas and depends on magic knowledge.

Anywhere that we're casting from a base class to a more specific one, I'd question why it's necessary in the first place. I'd wager in many of those cases, we either have access to a more specific pointer, or we should be retaining one instead.

In the cases where static_cast is being used properly or in an acceptable fashion, then yes. That would require a much more thorough evaluation of all these call sites though.

If we think that is worth the additional effort, I can do so.

PatTheMav replied:
Much appreciated. It's not a hard requirement, but in the interest of being better about documenting foot guns and potentially dangerous implementations it's appreciated.

PatTheMav commented on frontend/settings/OBSBasicSettings.cpp:
```diff
@@ -314,7 +314,7 @@ void RestrictResetBitrates(initializer_list<QComboBox *> boxes, int maxbitrate);
 OBSBasicSettings::OBSBasicSettings(QWidget *parent)
 	: QDialog(parent),
-	  main(qobject_cast<OBSBasic *>(parent)),
+	  main(static_cast<OBSBasic *>(parent)),
```
This change will allow creation of `OBSBasicSettings` objects with arbitrary `QWidget`s as parent pointers, leading to potential crashes at arbitrary points when `main` is interacted with in a way that is only possible with an actual `OBSBasic` instance.

The `qobject_cast` could at least ensure that `main` is a `nullptr`, but then the constructor body should actually check if `main` is not nulled and throw an exception otherwise. Instead it currently calls `EnableOutputs` unconditionally.

So if it is indeed required that `OBSBasicSettings` can only ever be instantiated with an `OBSBasic` parent, that should be explicitly communicated by the API instead of accepting an arbitrary `QWidget` pointer. This is a class of errors and constraints that we can and should enforce at compile time.

Warchamp7 replied:
This constructor very likely predates the existence of `OBSBasic::Get()` so I will remove this cast entirely.

PatTheMav commented on frontend/settings/OBSBasicSettings_Stream.cpp:
```diff
@@ -787,7 +787,7 @@ OBSService OBSBasicSettings::SpawnTempService()
 void OBSBasicSettings::OnOAuthStreamKeyConnected()
 {
-	OAuthStreamKey *a = reinterpret_cast<OAuthStreamKey *>(auth.get());
+	OAuthStreamKey *a = static_cast<OAuthStreamKey *>(auth.get());
```
I guess it's "fine" for now because every `OAuth` implementation in the program currently uses `OAuthStreamKey`.

But wouldn't it make more sene if the pointer is passed into the function directly, like so:

```C++
	if (type == Auth::Type::OAuth_StreamKey || type == Auth::Type::OAuth_LinkedAccount) {
        OAuthStreamKey *streamKey = static_cast<OAuthStreamKey *>(auth.get());

		OnOAuthStreamKeyConnected(streamKey);
	}
```

At least `OBSBasicSettings::OnAuthConnected` runs _some_ checks and balances here, but it also requires that whatever `ui->service->currentText()` returns as the UI label for the service selection is already in sync with the global `auth` object (again there is no other subclass of it yet, but as designed there could be). 🫠

PatTheMav reviewed (commented):
Even though it might seem superfluous to use `qobject_cast` or `dynamic_cast` without checking the result for `nullptr` after,  even a missing `nullptr` check will create deterministic crash points (mainly when that pointer is dereferenced).

But as the crash by `UpdateContextBar` demonstrates, `static_cast` will produce a valid pointer and thus dereferencing will succeed and if both types happen to have identical member or method names (like `GetSource`) code might potentially succeed and produce corrupted data or crash at arbitrary points later.

Otherwise I agree that where no checks and balances are used (and code is pretty blasé about safety so far anyway) using `static_cast` directly seems more "honest".

Warchamp7 committed:
squashme: Add fixme notes

Warchamp7 committed:
squashme: Fix source tree model cast

SourceTree is designed for SourceTreeModel and should avoid
using any other model. Enforce this on the SourceTree itself
